exonum/src/node/connect_list.rs

```diff
@@ -0,0 +1,239 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Module that provides methods for managing the mapping between peers
+//! public keys and IP-addresses.
+
+use std::collections::{BTreeMap, BTreeSet};
+use std::net::SocketAddr;
+
+use blockchain::ValidatorKeys;
+use crypto::PublicKey;
+use helpers::fabric::NodePublicConfig;
+use node::ConnectInfo;
+
+// TODO: Don't reload whitelisted_peers if path the same. (ECR-172)
+
+/// `ConnectList` stores mapping between IP-addresses and public keys.
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+pub struct ConnectList {
+    /// Peers to which we can connect.
+    #[serde(default)]
+    pub peers: BTreeMap<PublicKey, SocketAddr>,
+}
+
+impl ConnectList {
+    /// Returns `true` if a peer with the given public key can connect.
+    pub fn allow(&self, peer: &PublicKey) -> bool {
+        self.peers.contains_key(peer)
+    }
+
+    /// Adds peer to the ConnectList.
+    pub fn add(&mut self, peer: ConnectInfo) {
+        self.peers.insert(peer.public_key, peer.address);
+    }
+
+    /// Creates `ConnectList` from validators keys and corresponding IP addresses.
+    pub fn from_validator_keys(validators_keys: &[ValidatorKeys], peers: &[SocketAddr]) -> Self {
+        let peers: BTreeMap<PublicKey, SocketAddr> = peers
+            .iter()
+            .zip(validators_keys.iter())
+            .map(|(p, v)| (v.consensus_key, *p))
+            .collect();
+
+        ConnectList { peers }
+    }
+
+    /// Creates `ConnectList` from validators public configs.
+    pub fn from_node_config(list: &[NodePublicConfig]) -> Self {
+        let peers: BTreeMap<PublicKey, SocketAddr> = list.iter()
+            .map(|config| (config.validator_keys.consensus_key, config.addr))
+            .collect();
+
+        ConnectList { peers }
+    }
+
+    /// Check if we allow to connect to `address`.
+    pub fn address_allowed(&self, address: &SocketAddr) -> bool {
+        self.peers.values().any(|a| a == address)
+    }
+
+    /// Refresh `ConnectList` peers if validators has changed.
+    pub fn refresh(&mut self, validators_keys: &[ValidatorKeys]) {
```

Comment:
The method name / comment are non-descriptive; they don't imply that the list is strictly filtered.